### PR TITLE
gp_saml_gui.py: ensure that proxies are disabled

### PR DESCRIPTION
**Pull request by mchehab**
Using a proxy while trying to connect to VPN is problematic, as proxies are usually behind the VPN.

So, disable it when opening the WebKit window.

### CODE REVIEW

**Comment by dlenski** [comment]
> Using a proxy while trying to connect to VPN is problematic, as proxies are usually behind the VPN.

While that's a _common_ use case, it's also not unheard-of that VPN users _have to go through a proxy_ to access the VPN server. (It's a dumb and inefficient setup, but I've heard of it repeatedly with different corporate VPNs.)

Perhaps you could put this behind a `--no-proxy` option or similar?

File: gp_saml_gui.py [hunk 1]
```diff
@@ -65,6 +65,10 @@ def __init__(self, uri, html=None, verbose=False, cookies=None, verify=True, use
             self.cookies.set_persistent_storage(cookies, WebKit2.CookiePersistentStorage.TEXT)
         self.wview = WebKit2.WebView()
 
+	# Ensure that proxy will be disabled
+        data_manager = self.ctx.get_website_data_manager()
+        data_manager.set_network_proxy_settings(WebKit2.NetworkProxyMode.NO_PROXY, None)
```
Comment (dlenski) [inline gp_saml_gui.py:70]: I try to ensure that the network and TLS settings used for the webview are as closely matched as possible to the settings used when launching `openconnect` itself.

Could you please add `--no-proxy` to the OpenConnect command-line building (in `main`), whenever the proxy is disabled for the webview?

**Review by dlenski — CHANGES_REQUESTED** [review]

File: gp_saml_gui.py (hunk 1 above)
Comment (mchehab) [reply gp_saml_gui.py]: Sure. Just pushed a new version. While here:
- I opted to pass arguments as a dict to the class, as it makes easier to add more arguments;
- I also fixed an issue if one tries to close the login window (for example, when gp_saml_gui is called with an already-validated token).

File: gp_saml_gui.py [hunk 2]
```diff
@@ -124,6 +124,8 @@ def on_load_changed(self, webview, event):
         uri = mr.get_uri()
         rs = mr.get_response()
         h = rs.get_http_headers() if rs else None
+        if not h:
+            return
```
Comment (dlenski) [inline gp_saml_gui.py:133]: Could also just be…

```py
if not rs:
  return
h = rs.get_http_headers()
```

File: gp_saml_gui.py [hunk 3]
```diff
@@ -45,7 +45,8 @@ def handle_comment(self, data: str) -> None:
 
 
 class SAMLLoginView:
-    def __init__(self, uri, html=None, verbose=False, cookies=None, verify=True, user_agent=None):
+    def __init__(self, html, uri, args):
```
Comment (dlenski) [inline gp_saml_gui.py:48]: Wait… you have _reversed_ the order of the `html` and `uri` arguments in the function definition, but you _haven't_ reversed them in the callsite at L360.

How can this work?

Comment (mchehab) [reply gp_saml_gui.py]: heh, true. Fixed and tested:

```
writing RSA key
Looking for SAML auth tags in response to https://<vpn-url>/ssl-vpn/prelogin.esp...
Got SAML REDIRECT, opening browser...
[PAGE   ] Finished loading page https://login.microsoftonline.com/...
[PAGE   ] Finished loading page https://<vpn-url>/SAML20/SP/ACS
[SAML   ] Got SAML result headers: ...
[SAML   ] Got all required SAML headers, done.

SAML response converted to OpenConnect command line invocation:
...
SAML response converted to test-globalprotect-login.py invocation:
...
Launching OpenConnect with sudo, equivalent to:
...
Session authentication will expire at Fri Mar  8 23:17:00 2024

Using vhost-net for tun acceleration, ring size 32
```

**Comment by pboushy** [comment]
@dlenski - anything pending on this? We have several people who would benefit from this being added.

We're working on some other workarounds in the meantime, but this would go a long way to resolving.

File: gp_saml_gui.py (hunk 1 above)
Comment (mchehab) [reply gp_saml_gui.py]: Another patch with the same fix got merged:
```
c9e783116ede ("fix exception when checking about:blank response")
```
So, I dropped this patch: `a72825e87f18 ("Properly handle close window event without login")`.

Comment (mchehab) [reply gp_saml_gui.py]: > Could you please add `--no-proxy` to the OpenConnect command-line building (in `main`), whenever the proxy is disabled for the webview?
Added.

**Review by dlenski — APPROVED** [review]

**Pull request merged**